### PR TITLE
feat: Sales Order status for pick list

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
[Task](https://bloomstack.com/desk#Form/Task/TASK-2020-01422)

- [x] Add a percent field called % Picked (call it per_picked)
- [x] Write a patch to set the value based on how many items in the order have been picked (check existing Pick Lists against every order, and set the per_picked field as picked_qty / ordered_qty)
**On the Sales Order's listview (sales_order_list.js), add two statuses before "To Deliver" and "To Deliver and Bill":**
- [x] To Pick: with the conditions of per_picked > 1 and per_billed == 100
- [x] To Pick and Bill: with the conditions of per_picked > 1 and per_billed < 100

### CODE REVIEW

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import frappe
+
+def execute():
+	frappe.reload_doctype("Sales Order")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Use `reload_doc` instead with the `force` flag set to True. We've seen some issues with `reload_doctype`

File: erpnext/selling/doctype/customer/customer.json [hunk 2]
```diff
@@ -1,5 +1,4 @@
 {
```
Comment (Alchez) [inline erpnext/selling/doctype/customer/customer.json]: Do you need these changes in this PR?

File: erpnext/selling/doctype/sales_order/sales_order.json [hunk 3]
```diff
@@ -141,7 +141,8 @@
   "to_date",
   "column_break_108",
   "auto_repeat",
-  "update_auto_repeat_reference"
+  "update_auto_repeat_reference",
+  "is_completed"
```
Comment (Alchez) [inline erpnext/selling/doctype/sales_order/sales_order.json]: The definition for this field?

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 4]
```diff
@@ -15,18 +15,24 @@ frappe.listview_settings['Sales Order'] = {
 			// not delivered & overdue
 				return [__("Overdue"), "red",
 					"per_delivered,<,100|delivery_date,<,Today|status,!=,Closed"];
+			} else if (doc.docstatus == 1 && doc.per_picked == 100 && doc.per_delivered < 100) {
```
Comment (Alchez) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]:
```suggestion
			} else if (doc.docstatus == 1 && doc.per_billed == 100 && doc.per_delivered < 100) {
```

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 5]
```diff
@@ -15,18 +15,24 @@ frappe.listview_settings['Sales Order'] = {
...
+				return [__("To Pick"), "orange",
+					"per_picked,<,100|per_billed,=,100|status,!=,Closed"];
```
Comment (Alchez) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]: Add `per_delivered` here too

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+	frappe.reload_doctype("Pick List Item")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: I don't think you need to reload Pick List, there's no new fields there

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+	sales_orders = frappe.get_all("Sales Order", fields=["name"])
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: This can just be:

```suggestion
	sales_orders = frappe.get_all("Sales Order")
```

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+	for order in sales_orders:
+		pick_list_items = frappe.get_all("Pick List Item", filters={"sales_order":order.name}, fields= ["sum(qty) as total_qty", "sum(picked_qty) as picked_qty"], group_by="item_code")
+		picked_qty = sum(d['picked_qty'] for d in pick_list_items)
+		ordered_qty = sum(d['total_qty'] for d in pick_list_items)
+
+		if picked_qty and ordered_qty:
+			per_picked = (picked_qty / ordered_qty) * 100
+			frappe.db.set_value("Sales Order", order.name, "per_picked", per_picked)
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Does this update the status accordingly?

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/customer/customer.json (hunk 2 above)
Comment (MyuddinKhatri) [reply erpnext/selling/doctype/customer/customer.json]: No

File: erpnext/selling/doctype/sales_order/sales_order_list.js (hunk 5 above)
Comment (MyuddinKhatri) [reply erpnext/selling/doctype/sales_order/sales_order_list.js]: per_delivered or per_billed?

Comment (Alchez) [reply erpnext/selling/doctype/sales_order/sales_order_list.js]: On lines 19-20, I mean. It won't add the `per_delivered` filter

Comment (MyuddinKhatri) [reply erpnext/selling/doctype/sales_order/sales_order_list.js]: okay understood

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import frappe
+
+def execute():
+	frappe.reload_doc('selling', 'doctype', 'sales_order', force=True)
+	sales_orders = frappe.get_all("Sales Order")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Only get submitted orders

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+	for order in sales_orders:
+		pick_list_items = frappe.get_all("Pick List Item", filters={"sales_order":order.name}, fields= ["sum(qty) as total_qty", "sum(picked_qty) as picked_qty"], group_by="item_code")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Only get submitted pick lists:

```suggestion
		pick_list_items = frappe.get_all("Pick List Item",
			filters={"docstatus": 1, "sales_order": order.name},
			fields= ["sum(qty) as total_qty", "sum(picked_qty) as picked_qty"],
			group_by="item_code")
```

**Review by Alchez — CHANGES_REQUESTED** [review]

**Comment by Alchez** [comment]
@MyuddinKhatri please check Travis tests as well

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py (hunk 8 above)
Comment (Alchez) [reply erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: @MyuddinKhatri, I tested this out locally with a recent Humboldt backup, and old orders are not being set to "To Pick" or "To Pick and Bill", wherever applicable. You might want to handle that in the patch too.

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order_list.js (hunk 4 above)
Comment (Alchez) [reply erpnext/selling/doctype/sales_order/sales_order_list.js]: @MyuddinKhatri, I think you missed this one

File: erpnext/selling/doctype/sales_order/sales_order_list.js (hunk 5 above)
Comment (Alchez) [reply erpnext/selling/doctype/sales_order/sales_order_list.js]: @MyuddinKhatri, I think you missed this one

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 11]
```diff
@@ -15,18 +15,24 @@ frappe.listview_settings['Sales Order'] = {
 			// not delivered & overdue
 				return [__("Overdue"), "red",
 					"per_delivered,<,100|delivery_date,<,Today|status,!=,Closed"];
+			} else if (doc.docstatus == 1 && doc.per_billed == 100 && doc.per_delivered < 100) {
```
Comment (Alchez) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]: Change the `per_delivered` to `per_picked`. The `per_delivered` is already happening on line 13. Also do this on line 21.

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+			fields= ["sum(qty) as total_qty", "sum(picked_qty) as picked_qty"],
+			group_by="item_code")
+		picked_qty = sum(d['picked_qty'] for d in pick_list_items)
+		ordered_qty = sum(d['total_qty'] for d in pick_list_items)
+
+		if picked_qty and ordered_qty:
+			per_picked = (picked_qty / ordered_qty) * 100
+			frappe.db.set_value("Sales Order", order.name, "per_picked", per_picked)
+			if per_picked < 100 and order.per_delivered < 100 and order.per_billed == 100:
+				frappe.db.set_value("Sales Order", order.name, "status", "To Pick")
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Mark these to not have their modified times updated, since we don't want to push them up to the top of the list view and make it seem like they were triggered by a user action.

**Review by Alchez — CHANGES_REQUESTED** [review]

**Review by bhattdevarsh — DISMISSED** [review]

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+def execute():
+	frappe.reload_doc('selling', 'doctype', 'sales_order', force=True)
+	frappe.reload_doc('stock', 'doctype', 'pick_list_item', force=True)
+	sales_orders = frappe.get_all("Sales Order",
+		filters={"docstatus": 1},
+		fields=["per_delivered", "per_billed"])
+
+	for order in sales_orders:
+		pick_list_items = frappe.get_all("Pick List Item",
+			filters={"docstatus": 1, "sales_order": order.name},
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py:15]: You're not getting the `name` field in the `get_all`, so none of the code below this will work.

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+			fields= ["sum(qty) as total_qty", "sum(picked_qty) as picked_qty"],
+			group_by="item_code")
+		picked_qty = sum(d['picked_qty'] for d in pick_list_items)
+		ordered_qty = sum(d['total_qty'] for d in pick_list_items)
+
+		if ordered_qty:
+			per_picked = (picked_qty / ordered_qty) * 100
+			frappe.db.set_value("Sales Order", order.name, "per_picked", per_picked)
```
Comment (Alchez) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py]: Add an `update_modified` here too

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 15]
```diff
@@ -182,6 +185,11 @@ def update_package_tag(self, reset=False):
 					})
 					package_tag.save()
 
+	def set_per_picked(self):
+		picked_qty = sum(d.picked_qty for d in self.locations)
+		ordered_qty = sum(d.qty for d in self.locations)
+		per_picked = (picked_qty / ordered_qty) * 100
+		frappe.db.set_value("Sales Order", self.locations[0].sales_order, "per_picked", per_picked)
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: Avoid doing this using `set_value`, set it in the doc and save it instead.

Also, Humboldt's flow usually uses only 1 order, but our system should be able to handle multiple orders.

Break this down as:

```python
for order in all_orders:
	picked_qty = sum(picked_qty)
	ordered_qty = sum(ordered_qty)
	per_picked = (picked_qty / ordered_qty) * 100

	order = frappe.get_doc(order)
	order.per_picked = per_picked
	order.save()
```

**Review by Alchez — CHANGES_REQUESTED** [review]

**Review by neilLasrado — APPROVED** [review]

**Comment by Alchez** [comment]
@neilLasrado, this seems fine to me for Bloomstack. See if this could cause issues for JHAudio's workflow.

**Comment by neilLasrado** [comment]
@Alchez we could add a setting in delivery settings to mandate picklist before delivery

**Comment by Alchez** [comment]
@neilLasrado, @MyuddinKhatri 

Sounds like a plan. We can start with a global toggle, and then later think about doing a per-order toggle.

**Review by nmimsnikhil — DISMISSED** [review]

**Comment by neilLasrado** [comment]
@MyuddinKhatri @Alchez did we add a setting in delivery settings ? Also, can you resolve the merge conflicts and push ?

**Comment by neilLasrado** [comment]
@MyuddinKhatri Can you make the changes, Also fix the merge conflicts.

**Comment by neilLasrado** [comment]
@MyuddinKhatri Can you please fix this PR?

**Review by neilLasrado — DISMISSED** [review]

**Comment by neilLasrado** [comment]
@MyuddinKhatri I cant see the use Pick List checkbox in the delivery settings. Have you added that?

**Comment by MyuddinKhatri** [comment]
> @MyuddinKhatri I cant see the use Pick List checkbox in the delivery settings. Have you added that?

No, I will be doing it today.

**Review by neilLasrado — DISMISSED** [review]

File: erpnext/selling/doctype/sales_order/sales_order.json [hunk 16]
```diff
@@ -1240,12 +1248,18 @@
    "fieldname": "signature",
    "fieldtype": "Signature",
    "label": "Signature"
+  },
+  {
+   "default": "0",
+   "fieldname": "ignore_pick_list",
```
Comment (neilLasrado) [inline erpnext/selling/doctype/sales_order/sales_order.json]: why did you add this?

Comment (MyuddinKhatri) [reply erpnext/selling/doctype/sales_order/sales_order.json]: To ignore pick list status if checked then it set statuses the old way.

**Comment by MyuddinKhatri** [comment]
@neilLasrado can you please have a look at this?

**Comment by neilLasrado** [comment]
@MyuddinKhatri did you work on this ?

**Comment by MyuddinKhatri** [comment]
@neilLasrado I have pushed the changes please have a look.

File: erpnext/controllers/status_updater.py [hunk 17]
```diff
@@ -36,6 +36,8 @@ def validate_status(status, options):
 	],
 	"Sales Order": [
 		["Draft", None],
+		["To Pick", "eval:self.per_picked > 1 and self.per_delivered < 100 and self.per_billed == 100 and self.docstatus == 1"],
```
Comment (neilLasrado) [inline erpnext/controllers/status_updater.py]: self.per_picked should be > 0 not 1. In case you have the per_picked as 0.5 this will fail.

**Review by neilLasrado — CHANGES_REQUESTED** [review]

**Review by neilLasrado — APPROVED** [review]

File: erpnext/patches/v13_0/set_per_picked_in_sales_order.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+import frappe
+
+def execute():
```
Comment (rvpasari) [inline erpnext/patches/v13_0/set_per_picked_in_sales_order.py:3]: Please add a docstring and comments on the functions

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 19]
```diff
@@ -192,6 +195,14 @@ def update_package_tag(self, reset=False):
 					})
 					package_tag.save()
 
+	def set_per_picked(self):
```
Comment (rvpasari) [inline erpnext/stock/doctype/pick_list/pick_list.py:198]: Please add docstrings and comments

**Review by neilLasrado — DISMISSED** [review]

**Comment by neilLasrado** [comment]
@MyuddinKhatri can you please add docstrings, rest of the PR is fine.
